bundles/io/org.eclipse.smarthome.io.rest/src/main/java/org/eclipse/smarthome/io/rest/Stream2JSONInputStream.java

```diff
@@ -0,0 +1,116 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.io.rest;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Iterator;
+import java.util.stream.Stream;
+
+import org.apache.commons.io.IOUtils;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+/**
+ * This {@link InputStream} will stream {@link Stream}s as JSON one item at a time. This will reduce memory usage when
+ * streaming large collections through the REST interface. The input stream creates one JSON representation at a time
+ * from the top level elements of the stream. For best performance a flattened stream should be provided. Otherwise a
+ * nested collections JSON representation will be fully transformed into memory.
+ *
+ * @author Henning Treu - initial contribution
+ *
+ */
+public class Stream2JSONInputStream extends InputStream {
+
+    private final Iterator<String> iterator;
+
+    private InputStream jsonElementStream;
+
+    private boolean firstIteratorElement;
+
+    private Gson GSON = new GsonBuilder().create();
+
+    /**
+     * Creates a new {@link Stream2JSONInputStream} backed by the given {@link Stream} source.
+     *
+     * @param source the {@link Stream} backing this input stream. Must not be null.
+     *
+     * @throws IllegalArgumentException in case the source is null.
+     */
+    public Stream2JSONInputStream(Stream<?> source) {
+        if (source == null) {
+            throw new IllegalArgumentException("The source must not be null!");
+        }
+
+        iterator = source.map(e -> GSON.toJson(e)).iterator();
+        jsonElementStream = IOUtils.toInputStream("");
+        firstIteratorElement = true;
+    }
+
+    @Override
+    public int read() throws IOException {
+        int result = jsonElementStream.read();
+
+        if (result == -1) { // the current JSON element was completely streamed
+            if (finished()) { // we are done streaming the collection
+                return -1;
+            }
+
+            fillBuffer(); // get the next element into the jsonInputStream
+            result = jsonElementStream.read();
+        }
+
+        return result;
+    }
+
+    @Override
+    public void close() throws IOException {
+        closeJSONStream();
+        super.close();
+    }
+
+    private void fillBuffer() throws IOException {
+        String prefix;
+        if (firstIteratorElement) {
+            prefix = "[";
+            firstIteratorElement = false;
+        } else {
+            prefix = ",";
+        }
+
+        String entity = iterator.hasNext() ? iterator.next() : "";
+
+        String postfix = "";
+        if (!iterator.hasNext()) {
+            postfix = "]";
+        }
+
+        try {
+            closeJSONStream();
+            jsonElementStream = IOUtils.toInputStream(prefix + entity + postfix, StandardCharsets.UTF_8.name());
```

Comment:
Shouldn't you use a separate try-catch-block for the closeJSONStream method call?

If the close fails there is nothing you can do (and IMHO) this is the reason why you ignore it silently.

If the `IOUtils.toInputStream` fails, then it should not be catched because it is a failure the caller needs to know.